### PR TITLE
Refactor semantics to allow expressions inside on{} blocks

### PR DESCRIPTION
**Pull request by jlu5**
(Sample outputs are not changed yet, that will be in a follow up)

The idea of this change is to allow nesting more complicated logic inside of `on{}` triggers, so that more complex code generated with recursion can output to one MC function instead of being split into many. This has fairly obvious impacts on output readability and (potentially) runtime performance too.

### CODE REVIEW

File: datapack.g4 [hunk 1]
```diff
@@ -54,19 +54,20 @@ primary_expression:
 	'(' expression ')'
 	| '[' (expression (',' expression)*)? ']' // list constructor
 	| '{' expression+ '}' // begin
-	| 'on' '(' trigger ')' '{' command* '}'
+	| 'on' '(' trigger ')' '{' expression* '}'
```
Comment (JustinHuPrime) [inline datapack.g4:57]: This shouldn't need any grammar changes

File: semantics.md [hunk 2]
```diff
@@ -64,7 +64,7 @@ Square brackets form a list constructor, which expects zero or more expressions
 
 Braces form a begin-expression, which evaluates all but the last expression for side effects only and produces the last expression's value.
 
-An on-expression, as a side effect, creates an internal-use advancement that, when triggered (according to the expression's trigger), evalues the given commands. It produces the namespaced id of that internal-use advancement.
+An on-expression, as a side effect, creates an internal-use advancement that, when triggered (according to the expression's trigger), evaluates the expressions defined inside of it. It produces the namespaced id of that internal-use advancement.
```
Comment (JustinHuPrime) [inline semantics.md:67]: I think this just needs modification to the literal command semantics - a literal command can either be a string or a list of strings - if it's a list, it's expanded as if multiple strings were given

Comment (JustinHuPrime) [reply semantics.md:67]: Then again - this is probably just clarifying my bad wording

File: semantics.md [hunk 3]
```diff
@@ -76,24 +76,30 @@ String literals exist, and use curly braces for string interpolation. They may a
...
 
-A literal command expects a string, and is inserted verbatim. No type checks happen.
+The typechecked triggers (currently `consume_item` and `inventory_changed`) take item specifications - we'd like to typecheck that the item or tag are valid items, but we can't do that without listing all the items and tags from Minecraft.
 
-## Triggers
+The free-form trigger takes a Minecraft trigger name - we'd like to typecheck that it's a valid name, but we can't do that without listing all of the triggers from Minecraft. This also doesn't allow parameters on the trigger.
 
-The load trigger is special, as it can't be combined at all. The trigger will fire when the datapack is reloaded or the server is started. (This is implemented using function tags.)
+`on` blocks cannot be nested.
```
Comment (JustinHuPrime) [inline semantics.md:91]: Yes they can
```
on (load) {
  {
    on (tick) {
      "say hi"
    }
    "say loaded"
  }
}

File: semantics.md [hunk 4]
```diff
@@ -76,24 +76,30 @@ String literals exist, and use curly braces for string interpolation. They may a
...
 
-The tick trigger is also special, as it can't be combined at all. The trigger will fire every single tick. (This is implemented using function tags.)
+### Commands
 
-Triggers can be combined together using an or - so the on-expression's commands will run when any trigger fires.
+Commands are triggered in Minecraft through the `run` statement, which allows for the following subcommands:
```
Comment (JustinHuPrime) [inline semantics.md:95]: Why are we adding this? We just need to modify raw commands to accept lists of strings. We could also make `grant`, `revoke`, and `execute` produce strings - just for consistency.

**Review by JustinHuPrime — CHANGES_REQUESTED** [review]

**Review by JustinHuPrime — APPROVED** [review]

**Review by kilowatt- — APPROVED** [review]

**Pull request merged**